src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1215,59 +1215,84 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
+                    resultType = TypeSymbolWithAnnotations.Create(getLeftResultType(leftResult.Type.TypeSymbol, rightResult.Type?.TypeSymbol), resultIsNullable);
+                    break;
+                case BoundNullCoalescingOperatorResultKind.LeftUnwrappedType:
+                    resultType = TypeSymbolWithAnnotations.Create(getLeftResultType(unwrapIfNullable(leftResult.Type.TypeSymbol), rightResult.Type?.TypeSymbol), resultIsNullable);
+                    break;
+                case BoundNullCoalescingOperatorResultKind.RightType:
+                    resultType = TypeSymbolWithAnnotations.Create(getRightResultType(leftResult.Type.TypeSymbol, rightResult.Type.TypeSymbol));
+                    break;
+                case BoundNullCoalescingOperatorResultKind.LeftUnwrappedRightType:
+                    resultType = TypeSymbolWithAnnotations.Create(getRightResultType(unwrapIfNullable(leftResult.Type.TypeSymbol), rightResult.Type.TypeSymbol));
+                    break;
+                case BoundNullCoalescingOperatorResultKind.RightDynamicType:
+                    resultType = TypeSymbolWithAnnotations.Create(rightResult.Type.TypeSymbol, resultIsNullable);
+                    break;
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(node.OperatorResultKind);
             }
-            else
-            {
-                // PROTOTYPE(NullableReferenceTypes): Conversions: NullCoalescingOperator
-#if DEBUG
-                //Debug.Assert((object)leftType == null || AreCloseEnough(leftType.TypeSymbol, node.Type));
-                //Debug.Assert((object)rightType == null || AreCloseEnough(rightType.TypeSymbol, node.Type));
-#endif
 
-                // PROTOTYPE(NullableReferenceTypes): Capture in BindNullCoalescingOperator
-                // which side provides type and use that to determine nullability.
-                resultType = TypeSymbolWithAnnotations.Create((leftType ?? rightType)?.TypeSymbol, isNullableIfReferenceType: rightType?.IsNullable & leftType?.IsNullable);
+            _result = resultType;
+            return null;
 
-                ReportNullabilityMismatchIfAny(node.LeftOperand, resultType, leftType);
-                ReportNullabilityMismatchIfAny(node.RightOperand, resultType, rightType);
+            bool? getIsNullable(BoundExpression e, Result r) => (r.Type is null) ? e.IsNullable() : r.Type.IsNullable;
+            TypeSymbol getLeftResultType(TypeSymbol leftType, TypeSymbol rightType)
+            {
+                // If there was an identity conversion between the two operands (in short, if
+                // there is no conversion on the right operand), then check nullable conversions
+                // in both directions since it's possible the right operand is the better result type.
+                if ((object)rightType != null &&
+                    node.RightOperand.Kind != BoundKind.Conversion &&
```

Comment:
>node.RightOperand.Kind != BoundKind.Conversion && [](start = 20, length = 49)

This condition feels suspicious. What if there is an explicit conversion on the right?

Reply:
If we are looking for identity we should be checking for identity rather than looking for some other clues. The RightConversion should be available on the node.

---
In reply to: [175210721](https://github.com/dotnet/roslyn/pull/25358#discussion_r175210721) [](ancestors = 175210721)

Reply:
Changed to check for `ExplicitCast`.

---
In reply to: [175211649](https://github.com/dotnet/roslyn/pull/25358#discussion_r175211649) [](ancestors = 175211649,175210721)